vscode URI incompatibilities

akosyakov opened the pull request:
fix #5308

akosyakov commented on packages/core/src/common/uri.ts:
```diff
@@ -14,9 +14,12 @@
 import Uri from 'vscode-uri';
 import { Path } from './path';
 
+setUriThrowOnMissingScheme(false);
```
We have to disable it because of plugin ext: https://github.com/theia-ide/theia/issues/4487

cc @evidolob

svenefftinge replied:
Please put your comment into the code

akosyakov replied:
done

akosyakov commented:
@marcdumais-work it would be good to review it that clients don't need to use yarn resolutions to work around vscode-uri runtime issues.

marcdumais-work commented:
I am trying to test this using a local NPM registry. I'll update when I have some results.

marcdumais-work commented:
@akosyakov I think you are on the right track. I built this PR and published the resulting Theia packages in a local registry. Then I used the registry to build an application, re-using the browser example package.json, with packages versions changed to what I published. 

The resulting Theia application starts. I played with it for a minute, and it seems to work ok. Here's the backend log (there was nothing of interest in the frontend log, that I saw): 

```
$ yarn theia start
yarn run v1.15.2
$ /opt/ericsson-dev/theia/examples/browser/b/node_modules/.bin/theia start
root INFO Theia app listening on http://localhost:3000.
root INFO Started watching the git repository: file:///opt/ericsson-dev/theia
root ERROR Uncaught Exception:  Error: Unhandled method getClientId
root ERROR Error: Request 'getClientId' failed
    at Proxy.<anonymous> (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/core/lib/common/messaging/proxy-factory.js:301:33)
    at HostedPluginProcessesCache.<anonymous> (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-processes-cache.js:85:57)
    at step (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-processes-cache.js:53:23)
    at Object.next (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-processes-cache.js:34:53)
    at /opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-processes-cache.js:28:71
    at new Promise (<anonymous>)
    at __awaiter (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-processes-cache.js:24:12)
    at HostedPluginProcessesCache.getLazyClientId (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-processes-cache.js:81:16)
    at HostedPluginProcess.<anonymous> (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-process.js:91:74)
    at step (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-process.js:56:23)
root INFO Using Git [2.21.0] from the PATH. (/usr/bin/git)
root INFO Checking whether '--no-optional-locks' can be used with the current Git executable. Minimum required version is '2.15.0'.
root INFO '--no-optional-locks' is a valid Git option for the current Git version: '2.21.0'.
root INFO [nsfw-watcher: 7680] Started watching: /home/lmcmcds/.theia
root INFO [nsfw-watcher: 7680] Started watching: /opt/ericsson-dev/theia/.theia/settings.json
root INFO [nsfw-watcher: 7680] Started watching: /opt/ericsson-dev/theia/.vscode/launch.json
root INFO [nsfw-watcher: 7680] Started watching: /opt/ericsson-dev/theia/.vscode/settings.json
root INFO [nsfw-watcher: 7680] Started watching: /opt/ericsson-dev/theia/packages/typescript/src/node/typescript-contribution.ts
root INFO [nsfw-watcher: 7680] Started watching: /opt/ericsson-dev/theia
root INFO [nsfw-watcher: 7680] Started watching: /opt/ericsson-dev/theia
root INFO [nsfw-watcher: 7680] Started watching: /opt/ericsson-dev/theia
root INFO Detected keyboard layout from browser API: US (PC)
root WARN Collided keybinding is ignored;  {"command":"outlineView:toggle","keybinding":"ctrlcmd+shift+s"}  collided with  {"command":"file.saveAs","keybinding":"ctrl+shift+s"}
root WARN Could not register keybinding:
  {"command":"outlineView:toggle","keybinding":"ctrlcmd+shift+s"}
Error: "ctrlcmd+shift+s" is in collision with something else [scope:0]
root WARN Collided keybinding is ignored;  {"command":"passthrough","keybinding":"ctrl+k","context":"terminalActive"}  collided with  {"command":"terminal:clear","keybinding":"ctrlcmd+k","context":"terminalActive"}
root WARN Could not register keybinding:
  {"command":"passthrough","keybinding":"ctrl+k","context":"terminalActive"}
Error: "ctrl+k" is in collision with something else [scope:0]
root WARN e.onStart is slow, took: 184.09500003326684 ms
root INFO Using Git [2.21.0] from the PATH. (/usr/bin/git)
root INFO [nsfw-watcher: 7797] Started watching: /home/lmcmcds/.theia
root WARN e.onStart is slow, took: 948.3449999243021 ms
root INFO [nsfw-watcher: 7797] Started watching: /opt/ericsson-dev/theia
root INFO Config file tasks.json does not exist under file:///opt/ericsson-dev/theia
root INFO [nsfw-watcher: 7797] Started watching: /opt/ericsson-dev/theia/.theia/settings.json
root INFO [nsfw-watcher: 7797] Started watching: /opt/ericsson-dev/theia/.vscode/launch.json
root INFO [nsfw-watcher: 7797] Started watching: /opt/ericsson-dev/theia/.vscode/settings.json
root INFO [nsfw-watcher: 7797] Started watching: /opt/ericsson-dev/theia
root INFO [nsfw-watcher: 7797] Started watching: /opt/ericsson-dev/theia
root INFO Stopped watching the git repository: file:///opt/ericsson-dev/theia
root INFO Started watching the git repository: file:///opt/ericsson-dev/theia
root WARN "tslint.enable" preference is not supported
root WARN "search.exclude" preference is not supported
root WARN "lcov.path" preference is not supported
root WARN "lcov.watch" preference is not supported
root WARN "files.insertFinalNewline" preference is not supported
```

akosyakov commented:
@marcdumais-work thx, i'm trying to fix ci

Can you reproduce this error reliably:
```
root ERROR Uncaught Exception:  Error: Unhandled method getClientId
root ERROR Error: Request 'getClientId' failed
    at Proxy.<anonymous> (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/core/lib/common/messaging/proxy-factory.js:301:33)
    at HostedPluginProcessesCache.<anonymous> (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-processes-cache.js:85:57)
    at step (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-processes-cache.js:53:23)
    at Object.next (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-processes-cache.js:34:53)
    at /opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-processes-cache.js:28:71
    at new Promise (<anonymous>)
    at __awaiter (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-processes-cache.js:24:12)
    at HostedPluginProcessesCache.getLazyClientId (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-processes-cache.js:81:16)
    at HostedPluginProcess.<anonymous> (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-process.js:91:74)
    at step (/opt/ericsson-dev/theia/examples/browser/b/node_modules/@theia/plugin-ext/lib/hosted/node/hosted-plugin-process.js:56:23)
```

Please open an issue if so.

akosyakov commented:
We have to fix #4487 first, otherwise `URI` silently fallback to `file` scheme.

akosyakov commented:
@svenefftinge i've removed `withoutScheme`, new vscode-uri does not allow URIs without them and silently use `file` scheme. It breaks since developers misused this method to get a path. I've refactored all places to get `path` properly.

akosyakov commented:
@marcdumais-work Could you please give a try again and if it is good merge it?

marcdumais-work commented:
> @marcdumais-work Could you please give a try again and if it is good merge it?

Will do -  have some errands to run; give me a few hours.

marcdumais-work commented:
@akosyakov 
```
> Can you reproduce this error reliably:

root ERROR Uncaught Exception:  Error: Unhandled method getClientId
[...]
```

I did not see this error with the latest version of this PR

marcdumais-work commented:
> Could you please give a try again and if it is good merge it?

I re-tested and there are a couple of errors/problems. But I see the same on master. So I'll go ahead and merge this PR, and check if there are existing issues for the problem I saw, and if not create new ones.

marcdumais-work commented:
@akosyakov Thanks for this fix!
